### DISCUSSION TITLE
Use e107::getTemplate inside shortcode

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5260] [General]
Is there anyway it's possible to use the e107::getTemplate function inside a shortcode sc_ definition code?
I tryed and it didn't worked...
To make it work i have to declare the $template variable already loaded in the main code as global.... 😢

**Comment by Vodhin** [comment]
Templates use shortcode to fill in the blanks and generate the HTML for the data shown. Shortcode should be used for gathering and parsing data. Shortcode files should have very little HTML only where needed (eg an element that needs an ID or data-attributes that you don't want to be lost when someone else makes or alters the template.

It is possible to do, but I think it is redundant. I started writing a long and complicated demo code but it got wiped out by accident. I started writing again and realized that it was giving me a headache (the code ends up recursive) - so what you are asking for is probably not a good practice.

I have just updated my estate plugin and you might want to peek at the following files:
```
estate_shortcodes.php
templates/estate_template.php (parses templates and tablerenders inside template file)
listings.php
estate_menu.php
```

Now to your problem:

`To make it work i have to declare the $template variable already loaded in the main code as global....` 

You can use the $GLOBALS variable inside the shortcode function, just be sure the files are called in the correct order in your whatever.php file. 

Let's say you have a template
```
$SPECIALCODE = '<div>{SOME_SHORTCODE}</div>';
$PLUGIN_TEMPLATE['view']['top'] = '<div>{GET_SOMETHING}</div>';
```

and in your shortcode:

```
function sc_get_something($parm){
    
    $tp = e107::getParser();
    return $tp->parseTemplate($GLOBALS['SPECIALCODE'], false, $GLOBALS['sc']);
    }

function sc_some_shortcode($parm){
    return $data; /// whatever it is;
    }

```

and finally in your whatever.php file:
```
$ns = e107::getRender();
$tp = e107::getParser();
$tmpl = e107::getTemplate('plugin_name');
$sc = e107::getScBatch('plugin_name', true); // <-- the template is available to shortcodes using $GLOBALS
$text = $tp->parseTemplate($tmpl['view']['top'], false, $sc);
$text .= $tp->parseTemplate($tmpl['view']['mid'], false, $sc);
$text .= $tp->parseTemplate($tmpl['view']['bot'], false, $sc);
$ns->tablerender('My Caption',$text);
```

You see, it is like trying to put the horse between two carts...

**Reply by rica-carv** [reply]
@Vodhin So for now the only chance is to have a global template if i want to use it in shortcode...

I need to use a template inside a shortcode, because i want to have all the "templating" inside a template file.
Imagine you got a shortcode that renders and displays a button, now imagine someone in it's template doen't want a button, just a simple tag....
I try always to be openminded on my plugin coding, and to give all the option to a designer, not to impose it what i want as a plugin writer, so my question.
I choose E107 because it's a lot customizable, not like the Wordpress "stone" approach, in where you are limited by their code. So i want to keep my plugins the same...

**Reply by Vodhin** [reply]
`Imagine you got a shortcode that renders and displays a button, now imagine someone in it's template doen't want a button, just a simple tag....`

trying to render a template inside a shortcode makes more work. Instead, code your shortcode to return several options using the $parm variable:

Shortcode:
```

  function sc_linkme($parm){
    if($parm['as']){
      switch($parm['as']){
        case 'url': 
          return $this->var['link_url'];
          break;
        case 'tag': 
          return '<a href="'.$this->var['link_url'].'" class="'.($parm['class'] ? $parm['class'] : 'link').'">'.$this->var['prop_name'].'</a>';
          break;
        case 'button': 
          return '<button class="btn '.($parm['class'] ? $parm['class'] : 'btn-default').'" onclick="window.location.assign(\''.$this->var['link_url'].'\')">'.$this->var['prop_name'].'</button>';
          break;
        }
      }
    else{
      // fallback if no $parm['as']
      return '<a href="'.$this->var['link_url'].'"'.($parm['class'] ? ' class="'.$parm['class'].'"' : '').'>'.$this->var['prop_name'].'</a>';
      }
    }
```
  
Template:
```
{LINKME:as=url} // returns only the url (or whatever data)
{LINKME:as=tag&class=superlink} // renders as <a  href="whatever" class="superlink"></a>
{LINKME:class=superlink} //also renders as <a href="whatever" class="superlink"></a>
{LINKMEk} //also renders as <a href="whatever"></a>

{LINKME:as=button&class=btn-primary} // renders as a <button class="btn btn-primary" onclick="whatever"></button>
{LINKME:as=button} // renders as a <button class="btn" onclick="whatever"></button>
```

Just be sure to leave a comment showing what parameters work

**Reply by rica-carv** [reply]
@Vodhin That will not work, because i have a recursive multidimensional array data to render inside the shortcode, so i'll need that template inside the shortcode. Globals are evil, so i ended up assigning the template to the shortcode varible, and call inside the shortcode. It works, butit's kind of crumbersome....
